src/core/exec-invoke.c

```diff
@@ -5431,6 +5472,45 @@ int exec_invoke(
         if (r < 0)
                 return r;
 
+        if (context->private_bpf) {
+                _cleanup_close_ int fs_fd = -EBADF, mnt_fd = -EBADF;
+
+                fs_fd = fsopen("bpf", 0);
+                if (fs_fd < 0) {
+                        *exit_status = EX_OSERR;
+                        return fs_fd;
+                }
+
+                r = send_one_fd(token_fds[0], fs_fd, 0);
+                if (r < 0) {
+                        *exit_status = EX_IOERR;
+                        return r;
+                }
+
+                r = wait_for_terminate_and_check("(bpffs)", TAKE_PIDREF(bpftoken_pid).pid, 0);
+                if (r < 0) {
+                        *exit_status = EX_OSERR;
+                        return r;
+                }
+                if (r != EXIT_SUCCESS) {
+                        /* If something strange happened with the child, let's consider this fatal, too */
+                        *exit_status = EX_OSERR;
+                        return -EIO;
+                }
+
+                mnt_fd = receive_one_fd(token_fds[0], 0);
+                if (mnt_fd < 0) {
+                        *exit_status = EX_IOERR;
+                        return fs_fd;
+                }
```

Comment:
All the above can be done after the first call to setup_private_users(). Please move the logic into a separate function that is called after setup_private_users().

Reply:
but setup_private_users() is done before setup_delegated_namespaces(), the fsopen() call must be done after all the unshares